scenario.py

```diff
@@ -189,10 +190,216 @@ def __init__(self, params, experiment_path):
+        partners_with_specific_clusters = [p for p in partners_list if p.cluster_split_option == 'specific']
+
+        # Compose the list of different labels in the dataset
+        labels = list(set(y_train))
+        random.seed(42)
+        random.shuffle(labels)
+
+        # Check coherence of the split option:
+        nb_diff_labels = len(labels)
+        sum_specific_clusters = sum([p.cluster_count for p in partners_with_specific_clusters])
+        if not partners_with_shared_clusters:
+            max_shared_clusters = 0
+        else:
+            max_shared_clusters = max([p.cluster_count for p in partners_with_shared_clusters])
+        assert sum_specific_clusters + max_shared_clusters  <= nb_diff_labels
+
+        # Stratify the dataset per labels
+        x_train_for_cluster, y_train_for_cluster, count_available_per_cluster = {}, {}, {}
+        for label in labels:
+            idx_in_full_trainset = np.where(y_train == label)
+            x_train_for_cluster[label] = x_train[idx_in_full_trainset]
+            y_train_for_cluster[label] = y_train[idx_in_full_trainset]
+            count_available_per_cluster[label] = len(y_train_for_cluster[label])
+
+        # Sort partners per nb of different clusters in descending order
+        partners_with_shared_clusters = sorted(
+            partners_with_shared_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+        partners_with_specific_clusters = sorted(
+            partners_with_specific_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+
+        # Initialize dict for storing which partners pick from which clusters
+        dict_specific, dict_shared = {}, {}
+
+        # First, partners getting data samples from clusters they don't share
+        position = 0
+        for p in partners_with_specific_clusters:
+            dict_specific[p.id] = {"clusters_list": labels[position:position + p.cluster_count]}
+            position += p.cluster_count
+
+        # Second, partners getting data samples from shared clusters
+        shared_clusters = labels[position:position + max_shared_clusters]
+        for p in partners_with_shared_clusters:
+            dict_shared[p.id] = {"clusters_list": random.sample(shared_clusters, k=p.cluster_count)}
+
+        # print(shared_clusters)  # DEBUG
+        # print(dict_specific)  # DEBUG
+        # print(dict_shared)  # DEBUG
+
+        # Compute the data amount limiting factor
+        resize_factor = 1
+
+        # For partners getting data samples from clusters they don't share
+        # ... compare the nb of available samples vs. the nb of samples initially configured
+        for p in partners_with_specific_clusters:
+            count_available = sum([count_available_per_cluster[cl] for cl in dict_specific[p.id]["clusters_list"]])
```

Comment:
cl means class ? So we can replace it by label

Reply:
`cl` here means cluster. I think it is clearer. What is shared are the clusters of data samples, not the labels. The labels are the names of the clusters and the criteria that defines them.